### PR TITLE
📝 Add FastAPI to Implementations

### PR DESCRIPTION
**Pull request by tiangolo**
This adds FastAPI to the ASGI implementations :tada: :rocket:

### CODE REVIEW

File: docs/implementations.rst [hunk 1]
```diff
@@ -65,3 +65,13 @@ Starlette
 
 Starlette is a minimalist ASGI library for writing against basic but powerful
 ``Request`` and ``Response`` classes. Supports HTTP.
+
+
+FastAPI
```
Comment (adamchainz) [inline docs/implementations.rst:70]: The list is in alphabetical order and I presume we'd want to keep it that way, so move above Quart

**Comment by tiangolo** [comment]
Not sure if you want to mention Starlette-based frameworks, that's your call :shrug: ...I can add them too if you want. I think Sanic also supports ASGI now, that could probably go in there too.

Would you like me to add them or would you prefer to keep only Starlette here?

**Comment by erm** [comment]
There was some discussion [here](https://github.com/django/asgiref/issues/110) touching on this topic more broadly. I think this PR is revealing how difficult it may be to sustain the implementation list in its current form, specifically the lack of a definitive criteria for inclusion in this case.

**Comment by tiangolo** [comment]
Thanks @erm !

It makes sense, in that case, I guess it's probably fine to just close this issue.

And maybe we could add a pointer to https://github.com/florimondmanca/awesome-asgi from here as it has many other related resources.

**Comment by tomchristie** [comment]
Personally I think it'd make sense to include FastAPI, Responder, Flama yeah - they're all significantly different from each other, and significantly different from working with Starlette as a framework, so it'd be a clear thumbs up from me. I also don't think they necessarily need to mention on this resource that they happen to use Starlette. (Tho it's always appreciated on their own docs when they do)

**Comment by adamchainz** [comment]
I think it might be more sustainable to remove the list here and point to awesome-asgi only, otherwise there will be two community lists of implementations.

**Comment by tiangolo** [comment]
I guess we can close this PR as I imagine those changes would be in a different one, right?

On the other hand, I'm not sure how would you like to structure it after removing the "implementations" section and where to put the link to awesome-asgi. I guess it might be simpler if you add it directly in a way you're comfortable with... otherwise, would you like me to propose something?

**Comment by andrewgodwin** [comment]
Let's at least add FastAPI in for now (could you fix the ordering so it matches the existing alphabetical ordering?)

I agree it would be nice to just have awesome-asgi, but we can do this quickly now, and not block small easy changes on a harder, bigger change.

**Comment by tiangolo** [comment]
Cool!

Sure, done :heavy_check_mark:

**Comment by andrewgodwin** [comment]
Thanks! We should definitely look at awesome-asgi too, though.

**Comment by Kitty2014** [comment]
@tiangolo  Whether FastAPI can support GRPC implementation?

**Comment by tiangolo** [comment]
@Kitty2014 it's probably better to ask it in an issue in the FastAPI repo to not pollute much here.

But the short answer is, not really.

FastAPI is designed for web servers/APIs, it is based on Starlette, Starlette is a web framework/toolkit, so it uses the ASGI sub-specs for HTTP and WebSockets. All that is handled and run by an ASGI server like Uvicorn that listens on the web ports and runs the application.


---

Now, about gRPC, ASGI:

GRPC would probably first need an ASGI sub-spec, then an implementation server. And then probably a framework for GRPC.

But gRPC already has all its own ideas, framework, and tools. If I remember correctly, the spec, communication, and everything are heavily based on protocol buffers, which in turn, are heavily based on a C++ implementation, that includes some quirks, for example, many object attributes can be both scalar values and sequences at the same time, and it heavily depends on classes (i.e. no `dict`s or `list`s) but without the benefit of classes, like static type checks, completion, etc, all that ends up being quite awkward in Python (I don't know if all that makes more sense in C++).

Also, if I remember correctly it's all quite tightly-coupled, and pretty much expects you to use their provided packages, so, being honest, if you have a simple problem and want to do gRPC you might be better off using those than through the whole route of the ASGI sub-spec. Of course, it would be totally feasible, but probably a lot more work.

---

Now, if what you want is a way to create some type of remote server that can easily communicate in a binary format with clients, then you can probably use FastAPI (or any other ASGI framework) with an ASGI middleware like https://github.com/florimondmanca/msgpack-asgi

**Pull request merged**